fennel/client_tests/test_discrete_window_aggregation.py

```diff
@@ -188,3 +137,89 @@ def pipeline(cls, event: Dataset):
         ),
     )
     assert df["count"].tolist() == [2, 3, 1, 1]
+
+
+@mock
+def test_keyed_discrete_hopping_window_aggregation(client):
+
+    data = pd.read_csv("fennel/client_tests/data/app_events.csv").assign(
+        timestamp=lambda x: pd.to_datetime(x["timestamp"], utc=True)
+    )
+
+    @source(webhook.endpoint("AppEvents"), cdc="append", disorder="14d")
+    @dataset
+    class AppEvents:
+        user_id: int
+        page_id: int
+        timestamp: datetime
+
+    @dataset(index=True)
+    class EventCount:
+        user_id: int = field(key=True)
+        window: Window = field(key=True)
+        count: int
+        max_page_id: int
+        timestamp: datetime
+
+        @pipeline
+        @inputs(AppEvents)
+        def pipeline(cls, event: Dataset):
+            return event.groupby(
+                "user_id", window=Hopping("1d", "1h")
+            ).aggregate(
+                count=Count(),
+                max_page_id=Max(of="page_id", default=0),
+                emit="final",
+            )
+
+    @dataset(index=True)
+    class EventCountAggregate:
+        user_id: int = field(key=True)
+        avg: float
+        timestamp: datetime
+
+        @pipeline
+        @inputs(EventCount)
+        def pipeline(cls, event: Dataset):
+            return event.groupby("user_id").aggregate(
+                avg=Average(of="count", window=Continuous("forever")),
+            )
+
+    client.commit(
+        datasets=[AppEvents, EventCount, EventCountAggregate],
+        message="first_commit",
+    )
+    client.log("fennel_webhook", "AppEvents", data)
+
+    # Test online lookups
+    window = Hopping("1d", "1h")
+    df, _ = client.lookup(
+        EventCount,
+        keys=pd.DataFrame(
+            {
+                "user_id": [1, 2, 3],
+                "window": [
+                    bucketize(
+                        datetime(2023, 5, 2, 13, 0, tzinfo=timezone.utc), window
+                    ),
+                    bucketize(
+                        datetime(2023, 5, 3, 13, 0, tzinfo=timezone.utc), window
+                    ),
+                    bucketize(
+                        datetime(2023, 5, 5, 15, 0, tzinfo=timezone.utc), window
+                    ),
+                ],
+            }
+        ),
+    )
+    assert df["count"].tolist() == [2, 1, 1]
+
+    df, _ = client.lookup(
+        EventCountAggregate,
+        keys=pd.DataFrame({"user_id": [1, 2, 3]}),
+    )
+    assert df["avg"].tolist() == [
+        2.4545454545454546,
+        1.7560975609756098,
+        1.6488549618320612,
+    ]
```

Comment:
Typically better to do something like - 

     expected = [19.554, 55.1033, 105.286]
        assert df["total"].tolist() == pytest.approx(expected, abs=1e-3)

Reply:
changed.